### PR TITLE
Enable the root to send back a DAO-ACK

### PR DESCRIPTION
**Pull request by yatch**
I found that the root node did not respond to a DAO with K-bit on. This PR solves the issue.

To test this change, I used `regression-tests/12-rpl/05-rpl-up-and-down-routes.csc` with putting the followings into `code/project-conf.h`.

> #define RPL_CONF_DAO_ACK 1
> #define RPL_CONF_WITH_NON_STORING 1
> #undef RPL_CONF_WITH_STORING
> #define RPL_CONF_WITH_STORING 0
> #undef RPL_CONF_MOP
> #define RPL_CONF_MOP RPL_MOP_NON_STORING

### CODE REVIEW

File: core/net/rpl/rpl-ext-header.c [hunk 1]
```diff
@@ -206,8 +208,10 @@ rpl_srh_get_next_hop(uip_ipaddr_t *ipaddr)
     }
   }
 
-  if(uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
-      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) {
+  if((uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
+      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) ||
+     (default_instance->current_dag->rank == ROOT_RANK(default_instance) &&
+      rpl_ns_get_node(dag, &UIP_IP_BUF->destipaddr) != NULL)) {
```
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: This confuses me. Can you explain?

File: core/net/rpl/rpl-ext-header.c [hunk 2]
```diff
@@ -179,9 +179,11 @@ rpl_srh_get_next_hop(uip_ipaddr_t *ipaddr)
 {
   uint8_t *uip_next_hdr;
   int last_uip_ext_len = uip_ext_len;
+  rpl_dag_t *dag;
 
   uip_ext_len = 0;
   uip_next_hdr = &UIP_IP_BUF->proto;
+  dag = rpl_get_dag(&UIP_IP_BUF->destipaddr);
```
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: Seems wrong to me. The IPv6 destination here is the next hop in SRH. Why would there be root of a DAG?

**Comment by simonduq** [comment]
I couldn't see the relation between the code and the PR description on DAO-ACK and k bit, can you please explain? Thanks

File: core/net/rpl/rpl-ext-header.c (hunk 2 above)
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: My mistake sorry, `rpl_get_dag` works for any node in the DAG, not only the root.

File: core/net/rpl/rpl-ext-header.c [hunk 3]
```diff
@@ -378,20 +388,20 @@ insert_srh_header()
...
     node = node->parent;
     path_len++;
   }
 
-  if(((DEBUG) & DEBUG_PRINT) && node != NULL) {
-    rpl_ns_get_node_global_addr(&node_addr, node);
-    PRINTF("RPL: SRH Next Hop ");
-    PRINT6ADDR(&node_addr);
-    PRINTF("\n");
-  }
```
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: Due to the condition of `node != root_node` in `while` at line-373, `node` is identical to `root_node` at line-139. Then, you'll get not the address of the direct next hop but the root address in a debug message with "RPL: SRH Next Hop ".

Here is an example of the debug output. In this case, the node of ID:3 is the root and has "aaaa::203:3:3:3" as a global address.  You can see "aaaa::203:3:3:3" is printed as "SRH Next Hop" in the example.

> 00:00.073	ID:3	IPv6 addresses: aaaa::203:3:3:3
> 00:00.073	ID:3	fe80::203:3:3:3
> (snip)
> 00:11.349	ID:3	RPL: Packet going up, sender closer 0 (299 < 128)
> 00:11.349	ID:3	RPL: Rank OK
> 00:11.349	ID:3	RPL: SRH creating source routing header with destination aaaa::208:8:8:8 
> 00:11.349	ID:3	RPL: SRH Hop aaaa::207:7:7:7
> 00:11.349	ID:3	RPL: SRH Next Hop aaaa::203:3:3:3

File: core/net/rpl/rpl-ext-header.c (hunk 1 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: There are two cases where `rpl_srh_get_next_hop()` should convert the global address in a destination address field to a correspondent link-local address:
 - Case-1. Any node in a RPL network processes a packet which has a source routing header.
 - Case-2. The root node processes a packet (originally) destined for one of its children.

The first case is handled by the original code. The second one is what I tried to handle by this change. The point is that such a packet as described in Case-2 does does not have a source routing header. And 'rpl_srh_get_next_hop()' does nothing for a packet without the header.

The additional condition that is combined with the original one by `||` consists of two parts:

- 2-1. `default_instance->current_dag->rank == ROOT_RANK(default_instance)`
- 2-2. `rpl_ns_get_node(dag, &UIP_IP_BUF->destipaddr) != NULL`

2-1 determines if the node is the root node. 2-2 determines if the destination is in the RPL network. If a more strict check is required, we need to check if the return value from `rpl_ns_get_node()` has `root_node` as its parent.

**Comment by yatch** [comment]
Sure, let me explain it. Sorry for lack of explanation... And, the title of PR may be inappropriate.

When the root receives a DAO like shown below, it decided to send back to a DAO-ACK to the correspondent node since the DAO has K-bit on.

dao

This decision is made in `dao_ack_input_nonstroing()`, at line-889 of rpl-icmp6.c. Then, the process moves into `dao_ack_output()`.

The first barrier is here. `dao_ack_output()` builds the body of a DAO-ACK in `UIP_ICMP_PAYLOAD` as follows:

[rpl-icmp6.c:1085]
>   buffer = UIP_ICMP_PAYLOAD;
> 
>   buffer[0] = instance->instance_id;
>   buffer[1] = 0;
>   buffer[2] = sequence;
>   buffer[3] = 0;

`dao_ack_output()` is supposed to put the body just after the IPv6 header. However, it cannot do that.

`UIP_ICMP_PAYLOAD` is defined with `((unsigned char *)&uip_buf[uip_l2_l3_icmp_hdr_len])`. And `uip_l2_l3_icmp_hdr_len` is defined with `(UIP_LLH_LEN + UIP_IPH_LEN + uip_ext_len + UIP_ICMPH_LEN)`. Since the DAO has a hop-by-hop option and `uip_ext_len` is not cleared, `uip_ext_len` remains 8. Then, the body of the DAO-ACK locates 8 octets behind from the intended position.

The process goes into `insert_srh_header()` through `uip_icmp6_send()` and `rpl_insert_header()`. `insert_srh_header()` makes room for a source routing header by the following code, where `uip_l2_l3_hdr_len` is defined with `(UIP_LLH_LEN + UIP_IPH_LEN + uip_ext_len)`.

[rpl-ext-header.c:412]
>    /* Move existing ext headers and payload uip_ext_len further */
>   memmove(uip_buf + uip_l2_l3_hdr_len + ext_len,
>              uip_buf + uip_l2_l3_hdr_len, uip_len - UIP_IPH_LEN);
>   memset(uip_buf + uip_l2_l3_hdr_len, 0, ext_len);
> 
>   /* Insert source routing header */
>   UIP_RH_BUF->next = UIP_IP_BUF->proto;
>   UIP_IP_BUF->proto = UIP_PROTO_ROUTING;

Since `uip_ext_len` is 8, the source routing header is located 8 octets behind from the end of the IPv6 header, while "Next Header" field of the IPv6 header has `UIP_PROTO_ROUTING`. Then, this packet is handed to `tcpip_ipv6_output()`. To get a link-local address of the next-hop, `rpl_srh_get_next_hop()` is called. 

[tcpip.c:552]
> #if UIP_CONF_IPV6_RPL && RPL_WITH_NON_STORING
>     uip_ipaddr_t ipaddr;
>     /* Look for a RPL Source Route */
>     if(rpl_srh_get_next_hop(&ipaddr)) {                                                                                                                                 
>       nexthop = &ipaddr;
>     }
> #endif /* UIP_CONF_IPV6_RPL && RPL_WITH_NON_STORING */

The second barrier is at this point. `rpl_srh_get_next_hop()` returns 0. As a result, `nexthop` remains `NULL`. Since the root does not have a route for the global address in the destination address field, the packet is going to be discarded.

Take a look at `rpl_srh_get_next_hop()` shown below, where `UIP_RH_BUF` is defined with `((struct uip_routing_hdr *)&uip_buf[uip_l2_l3_hdr_len])`, `uip_l2_l3_hdr_len` is `(UIP_LLH_LEN + UIP_IPH_LEN + uip_ext_len)`.

[rpl-ext-header.c:177]
> int
> rpl_srh_get_next_hop(uip_ipaddr_t *ipaddr)
> {
>   uint8_t *uip_next_hdr;
>   int last_uip_ext_len = uip_ext_len; 
>   uip_ext_len = 0;
>   uip_next_hdr = &UIP_IP_BUF->proto;

`*uip_next_hdr` is `UIP_PROTO_ROUTING`. So, the next `while` block is skipped. 

[rpl-ext-header.c:186] 
>   /* Look for routing header */
>   while(uip_next_hdr != NULL && *uip_next_hdr != UIP_PROTO_ROUTING) { 
>(snip)
>   }

Then, we're going to get a link-local address of the next-hop.

[rpl-ext-header.c:209]
>   if(uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
>      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) {
>     /* Routing header found. The next hop should be already copied as the IPv6 destination
>      * address, via rpl_process_srh_header. We turn this address into a link-local to enable
>      * forwarding to next hop */
>     uip_ipaddr_copy(ipaddr, &UIP_IP_BUF->destipaddr);
>     uip_create_linklocal_prefix(ipaddr);
>     uip_ext_len = last_uip_ext_len;
>     return 1;
>   }

Now, `UIP_RH_BUF` points to a byte just after the IPv6 header since `uip_ext_len` is set with zero at the beginning of the function. However, `UIP_RH_BUF->routing_type` is not `RPL_RH_TYPE_SRH` (3). Actually, `UIP_RH_BUF->routing_type` is 0x63 (99) which is a value within the hop-by-hop option remaining in the memory.

[rpl-ext-header.c:220]
>   uip_ext_len = last_uip_ext_len;                                                                                                                                       
>   return 0;                      

Then, `rpl_srh_get_next_hop()` returns 0.

One of the changes in this PR is resetting `uip_ext_len` before constructing the DAO-ACK body in `dao_ack_output`. This solves the most part of the issue.

Another change is made to prevent from making a source routing header with 0 in the "Hdr Ext Len" field. Here is a sample of such a source routing header:

dao-ack

Finally, the root need to route a DAO-ACK packet whose destination is a global address of one of its children. The packet does not have any source routing header, but we need to have the correspondent link-local address in `nexthop` at line-557 of tcpip.c. This is done by adding some lines of code into  `rpl_srh_get_next_hop()`.

The other change is fixing the debug output by `insert_srh_header`.

Hope this help...

File: core/net/rpl/rpl-ext-header.c (hunk 3 above)
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: Right, sorry, my code was a leftover of a former version where the while stopped at `node->parent == root_node`.
In this case I don't see any need for the branch `if(((DEBUG) & DEBUG_PRINT))` anymore, even after your patch, do you agree?

File: core/net/rpl/rpl-ext-header.c (hunk 1 above)
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: Ah I see, this goes along with the other change below that suppresses SRH in case of a direct child. Makes sense. For a stricter check how about simply checking if the node returned by `rpl_ns_get_node()` has a parent field equal to the root?

File: core/net/rpl/rpl-ext-header.c [hunk 4]
```diff
@@ -370,6 +374,12 @@ insert_srh_header()
   /* For simplicity, we use cmpri = cmpre */
   cmpri = 15;
   cmpre = 15;
+
+  if(node != NULL && node == root_node) {
```
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: very minor comment: we already know that `root_node` is not null so I would only check `node == root_node`.

File: core/net/rpl/rpl-ext-header.c (hunk 1 above)
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: Would be great if you could add a comment (briefly) explaining the above

File: core/net/rpl/rpl-icmp6.c [hunk 5]
```diff
@@ -1082,6 +1082,7 @@ dao_ack_output(rpl_instance_t *instance, uip_ipaddr_t *dest, uint8_t sequence)
   PRINT6ADDR(dest);
   PRINTF("\n");
 
+  uip_ext_len = 0;
```
Comment (simonduq) [inline core/net/rpl/rpl-icmp6.c]: why is that needed here but not in the other `_output` functions?

Comment (simonduq) [inline core/net/rpl/rpl-icmp6.c]: ok am now reading your detailed PR explanation :)

**Comment by simonduq** [comment]
OK, I understand the `dao_ack_output` part, but I'm concerned with the fix. All other `_output` functions in the file work with the assumption that the uip buf is cleared when they are called. I would keep this assumption consistent. My suggestion is to call `uip_clear_buf` before calling `dao_ack_output`.

**Comment by simonduq** [comment]
As for the second problem you are tackling, do you agree it arises only after the (useful) modification you made, of not inserting SRH in the particular case of a direct child? Did you check if this DAO-ACK process also works with non-storing in the multi-hop case?

Thanks!

File: core/net/rpl/rpl-ext-header.c (hunk 3 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: I don't have any particular opinion on that, actually. I just kept it as it was. :-)

I thought you did not want to call `rpl_ns_get_node_global_addr()` unless `DEBUG_PRINT` is defined. If it is not the case, it would be better to modify the part with a simple `else` statement.

File: core/net/rpl/rpl-ext-header.c (hunk 1 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: I got it :+1:

File: core/net/rpl/rpl-ext-header.c (hunk 4 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: Ah, OK. Since I don't use an arrow (`->`) operation with `node` in that `if` block, we don't need a NULL-check for `node`. I agree that `node != NULL` is unnecessary.

File: core/net/rpl/rpl-ext-header.c (hunk 3 above)
Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: Right but with your patch `rpl_ns_get_node_global_addr` is not needed anymore there :)
IMo the whole block should just go away.

**Comment by yatch** [comment]
I like to go with the assumption that the uip buf is cleared when any `_output` function in rpl_icmp6.c is called. And, I didn't know that useful function, actually macro, `uip_clera_buf()`! Thant's nice. It definitely should be used instead of `uip_ext_len = 0`.

Speaking of the second problem, are you talking about this one?

> The second barrier is at this point. rpl_srh_get_next_hop() returns 0.

Without the change to put `uip_ext_len = 0` in `dao_ack_output()`, this always arises. After applying this change of clearing `uip_ext_len`, I agree what you said. The following shows packets which were captured only with the `uip_ext_len = 0` change to aa6cfd71852714c5288ca18ae92c4178eef00993. It works in terms of networking.

zero-length-srh

The highlighted packet has a SRH with length-0. It is sent out by the root node because It satisfies the following condition in `rpl_srh_get_next_hop` and `fe80::207:7:7:7 in `nexthop`.

>  if(uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
>      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) {

And, the last part:

> Did you check if this DAO-ACK process also works with non-storing in the multi-hop case?

Yes, I checked it with a multi-hop case. Here is a network generated with `12-rpl/05-rpl-up-and-down-routes.csc` which I used for testing this patch. The node with 3 is the root.

topo

This wireshark capture shows the DAO-ACK for a node 3-hop away from the root:

multihop

File: core/net/rpl/rpl-ext-header.c (hunk 3 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: Oh, I overlooked `rpl_ns_get_node_global_addr` staying at line-385! Which do you mean by `the whole block`, only the `else if` block or the entire `if ... else if ...` block? Either choice would be okay for me, anyway.

Comment (simonduq) [inline core/net/rpl/rpl-ext-header.c]: I would keep the first `if`, I find it useful for debug, but would remove the `else if` part.

**Comment by simonduq** [comment]
OK, I think we're on the same page.
As for where to clear the uip buffer, for consistency with the rest of the file and the stack, I think this PR should stick to clearing before calling the function.
I'm open to different designs but this should go in different PR, to contiki-os.

File: core/net/rpl/rpl-ext-header.c (hunk 3 above)
Comment (yatch) [inline core/net/rpl/rpl-ext-header.c]: Thanks, understood.

**Comment by yatch** [comment]
I totally agree to stick with the current design of contiki. :-)

**Comment by yatch** [comment]
I made some mistakes in the series of the commits... Please hold on.

File: core/net/rpl/rpl-icmp6.c [hunk 6]
```diff
@@ -887,6 +887,7 @@ dao_input_nonstoring(void)
   }
 
   if(flags & RPL_DAO_K_FLAG) {
+    uip_clear_buf();
     dao_ack_output(instance, &dao_sender_addr, sequence);
```
Comment (yatch) [inline core/net/rpl/rpl-icmp6.c]: There are two more places to call `dao_ack_output()` in rpl-icmp6.c. `dao_input_storing()` calls them.

In order to keep consistency in the source file, it may be nice to put `uip_clear_buf()` before calling each of them. However it works without calling `uip_clear_buf()` there in Storing mode. In Storing mode, a DAO is destined for a link-local address and does not have a Hop-by-Hop option. I did not see any issue on DAO-ACK in testing with Storing mode.

Comment (simonduq) [inline core/net/rpl/rpl-icmp6.c]: Yes I would definitely add it there as well, even though we don't see a problem now, a bug might arise with future modifications!

Comment (yatch) [inline core/net/rpl/rpl-icmp6.c]: That's great; I like it :+1:

Comment (yatch) [inline core/net/rpl/rpl-icmp6.c]: I thought you would add uip_clear_buf() there by yourself; but you might want me to do so....

I've added two uip_clear_buf()s to make sure to clear the uIP buffer before calling dao_ack_output().

**Comment by simonduq** [comment]
Thanks, this looks good to me, let me know if I should merge or if there are other changes you wanted to add first

**Comment by simonduq** [comment]
Thanks for the contribution, merging!
